File: Final/Code/Project_4/BaiToan4/graph_tree_conversion.py
from collections import defaultdict

def matrix_to_list(matrix):
    adj_list = [[] for _ in range(len(matrix))]
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            if matrix[i][j] != 0:
                adj_list[i].append((j, matrix[i][j]))
    return adj_list

def list_to_matrix(adj_list):
    n = len(adj_list)
    matrix = [[0]*n for _ in range(n)]
    for u in range(n):
        for v, w in adj_list[u]:
            matrix[u][v] = w
    return matrix

def list_to_map(adj_list):
    adj_map = [dict() for _ in range(len(adj_list))]
    for u in range(len(adj_list)):
        for v, w in adj_list[u]:
            adj_map[u][v] = w
    return adj_map

class TreeNode:
    def __init__(self, id):
        self.id = id
        self.children = []

class FCNSNode:
    def __init__(self, id):
        self.id = id
        self.first_child = None
        self.next_sibling = None

def array_to_fcns(parent):
    n = len(parent)
    nodes = [FCNSNode(i) for i in range(n)]
    root = None
    last_child = [None] * n

    for i in range(n):
        p = parent[i]
        if p == -1:
            root = nodes[i]
        else:
            if not nodes[p].first_child:
                nodes[p].first_child = nodes[i]
            else:
                last_child[p].next_sibling = nodes[i]
            last_child[p] = nodes[i]
    return root

File: Final/Code/Project_4/BaiToan4/graph_tree_conversion.cpp
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <list>

using namespace std;

struct Edge {
    int to;
    int weight; 
};

vector<vector<Edge>> matrixToList(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<Edge>> adjList(n);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            if (matrix[i][j] != 0)
                adjList[i].push_back({j, matrix[i][j]});
    return adjList;
}

vector<vector<int>> listToMatrix(const vector<vector<Edge>>& adjList) {
    int n = adjList.size();
    vector<vector<int>> matrix(n, vector<int>(n, 0));
    for (int u = 0; u < n; u++)
        for (auto edge : adjList[u])
            matrix[u][edge.to] = edge.weight;
    return matrix;
}

vector<map<int, int>> listToMap(const vector<vector<Edge>>& adjList) {
    vector<map<int, int>> adjMap(adjList.size());
    for (int u = 0; u < adjList.size(); ++u)
        for (auto e : adjList[u])
            adjMap[u][e.to] = e.weight;
    return adjMap;
}

struct TreeNode {
    int id;
    vector<TreeNode*> children;
};

struct FCNSNode {
    int id;
    FCNSNode* firstChild;
    FCNSNode* nextSibling;
};

FCNSNode* buildFCNS(const vector<int>& parent) {
    int n = parent.size();
    vector<FCNSNode*> nodes(n, nullptr);
    for (int i = 0; i < n; i++)
        nodes[i] = new FCNSNode{i, nullptr, nullptr};

    FCNSNode* root = nullptr;
    vector<FCNSNode*> lastChild(n, nullptr);

    for (int i = 0; i < n; i++) {
        int p = parent[i];
        if (p == -1) {
            root = nodes[i];
        } else {
            if (!nodes[p]->firstChild)
                nodes[p]->firstChild = nodes[i];
            else
                lastChild[p]->nextSibling = nodes[i];
            lastChild[p] = nodes[i];
        }
    }

    return root;
}